DDD-7.md

```diff
@@ -0,0 +1,153 @@
+
+As per [Quarkus extension guide](https://quarkus.io/guides/writing-extensions), the `EmbeddedEngine` should be usable for Quarkus extension even with current implementation.
+New implementation should allow seamless integration with Quarkus as well.
+Similar to gRPC, a PoC should be done as a follow-up task.
+
+### Testing
+
+Testsuite will be changed to use only `DebeziumEngine` API.
+Most of the tests which use `DebeziumEngine` inherit from `AbstractConnectorTest`, where a `DebeziumEngine` instance is created.
+`AbstractConnectorTest` will contain a protected method which will be responsible for creating `DebeziumEngine`.
+When switching to the new `DebeziumEngine` implementation, the switching will be done by adjusting only this single method.
+This would also allow changing the engine implementation in the specific tests if needed.
+
+In the future, if there is any such need, the testsuite can be parameterized with `DebeziumEngine` implementation as a parameter.
+This would allow us to run the testsuite against multiple engine implementations.
+However, in the near future we don't expect there will be any such need, so for now a dedicated method within `AbstractConnectorTest`  should be sufficient.
+
+As the existing testsuite is based on `EmbeddedEngine` implementation, which provides richer API than the `DebeziumEngine` interface, it's not possible to use  `DebeziumEngine` API exclusively in the whole testsuite.
+It's possible to use `DebeziumEngine` API in most of the cases or introduce some helper methods instead, except one case - `EmbeddedEngine#runWithTask()`.
+It exposes the engine Kafka Connect `SourceTask` for the testing.
+If we get rid of this method, we would lose some important tests.
+To preserve the ability to test the engine source task and at the same time not break encapsulation of `DebeziumEngine`, a new interface `TestingDebeziumEngine` will be introduced.
+It will belong to the test package of the `debezium-embedded` module.
+The interface should contain only one method, `runWithTask(Consumer<SourceTask> consumer)`:
+
+```java
+public interface TestingDebeziumEngine<T> extends DebeziumEngine<T> {
+    /**
+     * Run consumer function with engine task, e.g. in case of Kafka with {@link SourceTask}.
+     * Effectively expose engine internal task for testing.
+     */
+    void runWithTask(Consumer<SourceTask> consumer);
+}
+```
+
+In the future, we may add more methods to this interface if it would be beneficial from the testing point of view to expose any other `DebeziumEngine` internals or add convenient methods for testing.
+Number of such methods should be, however, kept as minimal as possible and methods should be added only for a very good reason, as the interface will force `DebeziumEngine` implementations to implement all these methods as well if the implementation should be tested with the Debezium testsuite.
```

Comment:
I'd like to see a reference to performance testing too.

Reply:
Sure, completely forgot to mention it. The plan is to use JHM benchmark similar to `debezium-microbenchmark-oracle` with [`SimpleSourceConnector`](https://github.com/debezium/debezium/blob/main/debezium-embedded/src/main/java/io/debezium/connector/simple/SimpleSourceConnector.java) or maybe add another implementation similar to this, more suitable for perf. testing (possibly with option to configure how large/complex generated records should be).